torch/csrc/jit/ir.cpp

```diff
@@ -413,88 +427,126 @@ void Graph::lint() const {
-    check_node(n);
-  }
+    void check_block(const Block *b) {
+      for (auto input : b->inputs()) {
+        check_value(input);
+        JIT_ASSERT(input->node()->kind_ == kParam);
+      }
 
-  JIT_ASSERT(output_->kind() == kReturn);
-  check_node(output_);
+      for (auto n : b->nodes()) {
+        JIT_ASSERT(n->kind_ != kParam);
+        JIT_ASSERT(n->kind_ != kReturn);
+        check_node(n);
+      }
 
-  for (auto kv : anticipated_uses) {
-    JIT_ASSERT(kv.second == -1);
-  }
+      JIT_ASSERT(b->output_->kind() == kReturn);
+      check_node(b->output_);
+
+      // all_nodes
+      // - inputs_, output_ and nodes_ are all included in all_nodes
+      // - all_nodes does not contain dead nodes??? (likely to be temporarily
+      // suspended).  Weaker: all_nodes contains all inputs and returns
+      // - only one return node???
+
+      node_set nodes_set(ALL_OF(b->nodes()));
+      node_set inputs_set {b->input_};
+      node_set output_set {b->output_};
+      // TODO: Make a more type safe std::includes wrapper which disallows use on
+      // non-ordered containers
+      JIT_ASSERT(std::includes(ALL_OF(all_nodes_set), ALL_OF(nodes_set)));
+      JIT_ASSERT(std::includes(ALL_OF(all_nodes_set), ALL_OF(inputs_set)));
+      JIT_ASSERT(std::includes(ALL_OF(all_nodes_set), ALL_OF(output_set)));
+
+      sum_set.insert(ALL_OF(nodes_set));
+      sum_set.insert(ALL_OF(inputs_set));
+      sum_set.insert(ALL_OF(output_set));
+    }
+    void check_graph() {
+      node_set all_nodes_set(ALL_OF(g.all_nodes)); // NB: all_nodes is *unordered*
 
-  // all_nodes
-  // - inputs_, output_ and nodes_ are all included in all_nodes
-  // - all_nodes does not contain dead nodes??? (likely to be temporarily
-  // suspended).  Weaker: all_nodes contains all inputs and returns
-  // - only one return node???
-
-  node_set all_nodes_set(ALL_OF(all_nodes)); // NB: all_nodes is *unordered*
-  node_set nodes_set(ALL_OF(nodes()));
-  node_set inputs_set {input_};
-  node_set output_set{output_};
-  // TODO: Make a more type safe std::includes wrapper which disallows use on
-  // non-ordered containers
-  JIT_ASSERT(std::includes(ALL_OF(all_nodes_set), ALL_OF(nodes_set)));
-  JIT_ASSERT(std::includes(ALL_OF(all_nodes_set), ALL_OF(inputs_set)));
-  JIT_ASSERT(std::includes(ALL_OF(all_nodes_set), ALL_OF(output_set)));
-
-  node_set sum_set;
-  sum_set.insert(ALL_OF(nodes_set));
-  sum_set.insert(ALL_OF(inputs_set));
-  sum_set.insert(ALL_OF(output_set));
-  JIT_ASSERT(std::includes(ALL_OF(sum_set), ALL_OF(all_nodes_set)));
-
-  // graph->stage() should be equal to max(node.stage for node in graph->nodes())
-  if (begin() == end()) {
-    JIT_ASSERT(stage() == 0);
-  } else {
-    JIT_ASSERT(stage() == rbegin()->stage());
-  }
+      check_block(g.block_);
+      for (auto kv : anticipated_uses) {
+        JIT_ASSERT(kv.second == -1);
+      }
+      // graph->stage() should be equal to max(node.stage for node in graph->nodes())
```

Comment:
We can get rid of the old school stages soon right? :)

Reply:
Nope. We still use them in `autodiff.cpp`

Reply:
Although all graphs outside of autodiff will only have a single stage I think

Reply:
What I mean is that, the next stage can simply be another block embedded inside the graph.

Reply:
Ah yes, that makes sense

Reply:
After we do more bringup on the new jit, we can consider re-working how stages are defined. We need to keep the functionality for autodiff at least, but I also want to see what we need to do to optimize perf in the new system and if that still involves stages.